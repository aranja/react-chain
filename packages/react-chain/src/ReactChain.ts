import createBase from './ReactChainBase'
import { ReactElement } from 'react'
import Session from './Session'
import { getChainForTarget, renderRecursively, render } from './SessionUtils'

export type RenderTarget =
  'browser' |
  'server'

export type WrapElement =
  (renderChildren: () => Promise<null | ReactElement<any>>, context: any) =>
    ReactElement<any> | Promise<ReactElement<any>>

export type WrapRender =
  (render: Function) =>
    void

export type Middleware =
  (session: Session) =>
    (void | WrapElement)

export class ReactChain {
  protected middlewareChain: Array<Middleware> = []

  chain(middleware?: Middleware) {
    if (typeof middleware !== 'function') {
      throw new Error('A react-chain middleware should be a function')
    }

    this.middlewareChain.push(middleware)

    return this
  }

  async getElement(session = new Session()): Promise<any> {
    if (session.__firstRender) {
      session.__elementChain = [createBase]
      this.middlewareChain.forEach(middleware => {
        const createElement = middleware(session)
        if (createElement) {
          session.__elementChain.push(createElement)
        }
      })
      session.__firstRender = false
    }

    let index = 0
    const next = (): Promise<any> => {
      const createElement = session.__elementChain[index++]
      const renderChildren = session.__elementChain[index]
        ? next
        : () => Promise.resolve(null)
      return Promise.resolve(createElement(renderChildren, session))
    }

    return await next()
  }

  renderBrowser(session: Session, onRender: () => void) {
    render(session, 'browser')(onRender)
  }

  renderServer(session: Session, onRender: Function) {
    let body = ''
    render(session, 'server')(() => {
      body = onRender()
    })
    return body
  }
}

export default function createReactChain() {
  return new ReactChain()
}
