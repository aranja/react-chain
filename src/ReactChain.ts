import createSession from './Session'
import { isValidElement, ReactElement } from 'react'
import { validateElementCreator, unfoldRender, renderElementChain } from './utils'
import { Middleware, Session, CreateElement } from './types'
export { Middleware, Session, CreateElement }

export class ReactChain {
  middlewareChain: Array<Middleware> = []

  use(middleware?: ReactElement<any> | Middleware) {
    if (middleware && isValidElement(middleware)) {
      middleware = ((element) => () => () => element)(middleware)
    }
    if (typeof middleware !== 'function') {
      throw new Error('A react-chain middleware should be a function')
    }

    this.middlewareChain.push(middleware)

    return this
  }

  // Deprecated
  chain(middleware?: ReactElement<any> | Middleware) {
    return this.use(middleware)
  }

  createSession = createSession

  async getElement(session?: Session): Promise<any> {
    if (session == null) {
      throw new Error('Missing session object.')
    }

    if (session.__firstRender) {
      this.middlewareChain.forEach(middleware => {
        let createElement: void | CreateElement
        if ((createElement = validateElementCreator(middleware(session)))) {
          session.__elementChain.push(createElement)
        }
      })

      session.__firstRender = false
    }

    return await renderElementChain(session.__elementChain)
  }

  async renderBrowser(session: Session, onRender: (element: ReactElement<any>) => any) {
    const element = await this.getElement(session)
    return unfoldRender(session, 'browser', () => onRender(element))
  }

  async renderServer(session: Session, onRender: (element: ReactElement<any>) => any) {
    const element = await this.getElement(session)
    return unfoldRender(session, 'server', () => onRender(element))
  }
}

export default function createReactChain() {
  return new ReactChain()
}
