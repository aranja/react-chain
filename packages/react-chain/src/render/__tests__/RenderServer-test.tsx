import renderServer, { ServerConfig } from '../RenderServer'
import * as React from 'react'
import createReactChain from '../../ReactChain'
import { ReactChain } from '../../ReactChain'

describe('renderServer()', () => {
  let app: ReactChain

  beforeEach(() => {
    app = createReactChain()
  })

  it('should throw if no Document is defined', () => {
    expect(() => {
      renderServer(app)
    }).toThrowErrorMatchingSnapshot()
  })

  it('should render a Document', async () => {
    let statusCode = -1
    let html = ''

    const config = {
      Document: () => <html>
      <body><h1>hello</h1></body>
      </html>,
    }
    const res = {
      send(data: string) {
        html = data
      },
      status(code: number) {
        statusCode = code
      },
    }

    await renderServer(app, config)({}, res)

    expect(statusCode).toBe(200)
    expect(html).toEqual(
      `<!doctype html><html><body><h1>hello</h1></body></html>`,
    )
  })

  it('should expose `assets`, `context`, `req` and `res`', async () => {
    let props: any = {}

    const config = {
      Document: (_props: any) => {
        props = _props
        return <div />
      },
    }

    await renderServer(app, config as ServerConfig<any>)({}, {
      status() {},
      send() {},
    })

    expect(props).toHaveProperty('assets')
    expect(props).toHaveProperty('context')
    expect(props.context).toHaveProperty('req')
    expect(props.context).toHaveProperty('res')
    expect(props.context).not.toHaveProperty('on')
  })

  it('should return a 500 on error', async () => {
    let statusCode = -1
    let html = ''

    const config = {
      Document: ({ title, description, children }: any) => (
        <div>
          <h1>{title}</h1>
          <p>{description}</p>
          <div>{children}</div>
        </div>
      ),
    }

    app.getElement = () => {
      throw new Error('Error message goes here...')
    }

    const res = {
      send(data: string) {
        html = data
      },
      status(code: number) {
        statusCode = code
      },
    }

    await renderServer(app, config as ServerConfig<any>)({}, res)

    expect(statusCode).toBe(500)
    expect(html).toMatchSnapshot()
  })

  it('should call next on error', async () => {
    const next = jest.fn()

    const config = {
      Document: () => <div />,
    }

    const res = {
      send() {},
      status() {},
    }

    app.getElement = () => {
      throw new Error('Error message goes here...')
    }

    await renderServer(app, config as ServerConfig<any>)({}, res, next)

    expect(next).toHaveBeenCalled()
  })
})
