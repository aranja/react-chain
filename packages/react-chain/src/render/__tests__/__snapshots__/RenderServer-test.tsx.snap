// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderServer() should return a 500 on error 1`] = `"<!doctype html><div><h1>Internal Server Error</h1><p>Error message goes here...</p><div>Error: Error message goes here...</div></div>"`;

exports[`renderServer() should throw if no Document is defined 1`] = `"renderServer requires a Document React component."`;
